Add Shoryuken support

phstc opened the pull request:
This PR ports [Honeybadger support for Shoryuken](https://github.com/honeybadger-io/honeybadger-ruby/pull/190) over to Airbrake. 

## TODO

- [x] Add tests
- [x] Notify [batches](https://github.com/phstc/shoryuken/wiki/Middleware#be-careful-with-batchable-workers)

phstc commented on lib/airbrake/shoryuken/error_handler.rb:
```diff
@@ -0,0 +1,30 @@
+      private
+
+      def notify_airbrake(exception, context)
+        return unless (notice = Airbrake.build_notice(exception, context))
```
Does `build_notice` support `context` being passed as an array?

kyrylo replied:
No, it must be a Hash.

phstc replied:
@kyrylo can I `{ contents: [] }` then?

kyrylo replied:
Yes, but it will be added to `params`:

```rb
notice[:params][:contents] #=> []
```

phstc replied:
@kyrylo Thanks! I think `notify_airbrake(e, body.is_a?(Array) ? { batch: body } : body)` will do then.

kyrylo commented:
>  I would like to know if you guys are open to it.

Absolutely, thanks for this!

phstc commented:
@kyrylo it's ready for review! 🍻

kyrylo commented on airbrake.gemspec:
```diff
@@ -17,7 +17,7 @@ prioritization of exceptions so that when errors occur, your team can quickly
 
 Additionally, this gem includes integrations with such popular libraries and
-frameworks as Rails, Sinatra, Resque, Sidekiq, Delayed Job, ActiveJob and many
+frameworks as Rails, Sinatra, Resque, Sidekiq, Delayed Job, Shoryuken, ActiveJob and many
```
Please wrap this line at 80 characters.

kyrylo commented on lib/airbrake/shoryuken/error_handler.rb:
```diff
@@ -0,0 +1,34 @@
+    class ErrorHandler
+      def call(_worker, _queue, _sqs_msg, body)
+        yield
+      rescue => e
```
Nitpick: we use `exception` for exception objects.

kyrylo commented on lib/airbrake/shoryuken/error_handler.rb:
```diff
@@ -0,0 +1,34 @@
+      private
+
+      def notice_context(body)
+        body.is_a?(Array) ? { batch: body } : { body: body }
```
I think it would be better if this check was on line 21:

```rb
notice[:params] = body.is_a?(Array) ? { batch: body } : { body: body }
```

This way we don't execute it when `notify_airbrake` returns early (and we so far we try to use this style in other integrations).

phstc replied:
>This way we don't execute it when notify_airbrake returns early (and we so far we try to use this style in other integrations).

@kyrylo I'm moving it to the `notify_airbrake` method, but it won't help with that ^, because it needs to be passed in the `build_notice` call. So even for the earliest it can return, `notice_context` will be called.

kyrylo commented on lib/airbrake/shoryuken/error_handler.rb:
```diff
@@ -0,0 +1,34 @@
+
+      def notify_airbrake(exception, context)
+        return unless (notice = Airbrake.build_notice(exception, context))
+
```
It would be good to set action/component. Example:
https://github.com/airbrake/airbrake/blob/c45bd6e0b4955dd0a0ad7fa87590383cc65fd8a4/lib/airbrake/sidekiq/error_handler.rb#L19

I'm not sure what the action would be in case of Shoryuken, so please let me know what you think suits best for this integration (you can check other integrations to see what we usually set).

phstc replied:
@kyrylo I dunno what to add as action. I can try the class name as the given example.

kyrylo commented on spec/unit/shoryuken/error_handler_spec.rb:
```diff
@@ -0,0 +1,53 @@
+module Airbrake
+  ##
+  # Provides tests for the Shoryuken integration.
+  module Shoryuken
```
Would you mind unwrapping this? We don't use this style in this project. I think `RSpec.describe Airbrake::Shoryuken::ErrorHandler` should suffice.

kyrylo commented on spec/unit/shoryuken/error_handler_spec.rb:
```diff
@@ -0,0 +1,53 @@
+require 'spec_helper'
+require 'airbrake/shoryuken/error_handler'
```
Are you sure this line is needed?

phstc replied:
@kyrylo it is, beause [it's not requiring the error handler unless Shoryuken defined](https://github.com/phstc/airbrake/blob/41dc8f28970694621764206edf26e1d31fbd1563/lib/airbrake.rb#L26). And to have Shoryuken defined, users need to explicit add Shoryuken in their Gemfiles.

kyrylo commented on spec/unit/shoryuken/error_handler_spec.rb:
```diff
@@ -0,0 +1,53 @@
+      context "when there's an error" do
+        it 'notifies' do
+          expect(Airbrake).to receive(:build_notice).with(error, body: body).
+            and_return(notice)
```
We use Webmock to test the library, so please avoid mocks at all costs :)

This is the equivalent of this test: https://github.com/airbrake/airbrake/blob/c45bd6e0b4955dd0a0ad7fa87590383cc65fd8a4/spec/unit/sidekiq/error_handler_spec.rb#L29

kyrylo commented on spec/unit/shoryuken/error_handler_spec.rb:
```diff
@@ -0,0 +1,53 @@
+        end
+      end
+
+      context "when there's no error" do
```
I think this scenario is unnecessary to test. Instead, it would be nice to test the scenario when Airbrake is not configured: https://github.com/airbrake/airbrake/blob/c45bd6e0b4955dd0a0ad7fa87590383cc65fd8a4/spec/unit/sidekiq/error_handler_spec.rb#L37

phstc commented on spec/unit/shoryuken/error_handler_spec.rb:
```diff
@@ -0,0 +1,64 @@
+
+  def wait_for_a_request_with_body(body)
+    wait_for(a_request(:post, endpoint).with(body: body)).to have_been_made.once
+  end
```
This should be in a shared example to DRY it out.

kyrylo replied:
Yeah, I agree that this doesn't look good, however we use this method in many spec files, so this change is rather a job for a different PR because it's quite involving.

phstc commented:
@kyrylo feedbacks addressed 🍻  could you review it again?

kyrylo commented on spec/unit/shoryuken/error_handler_spec.rb:
```diff
@@ -0,0 +1,64 @@
+    'https://airbrake.io/api/v3/projects/113743/notices?key=fd04e13d806a90f96614ad8e529b2822'
+  end
+
+  class FooWorker; end
```
To avoid defining a class in global scope (so we don't leak data from this test), I would use the following:

```rb
let(:worker) do
  Class.new do
    def to_s
      'FooWorker'
    end
  end.new
end
```

I know, looks unusual, but does the job :) If you have better ideas, I'm happy to see them, too.

phstc replied:
@kyrylo I don't find it unusual, actually my first attempt was with `Class.new`, the issue with that is because we use [`worker.class.to_s`](https://github.com/phstc/airbrake/blob/add-shoryuken-support/lib/airbrake/shoryuken/error_handler.rb#L21), which will return something like `Class:0x007fd03e57c360`.

Ideally, I think this should use [`worker_name`](https://github.com/phstc/shoryuken/blob/master/lib/shoryuken/util.rb#L40), but I didn't want to add Shoryuken as a development dependency in the gemspec.

WDYT?

> To avoid defining a class in global scope

If we go back with unwrapping the name space it won't be that global anymore.

kyrylo replied:
> the issue with that is because we use worker.class.to_s, which will return something like Class:0x007fd03e57c360

I can see that problem, so that's why I suggested to redefine `to_s`. If not, we can use a regexp to match that value.

> Ideally, I think this should use worker_name, but I didn't want to add Shoryuken as a development dependency in the gemspec.

Thanks for linking to that file. I'll be useful if eventually we'll come up with some integration tests in the future (say, with Rails). We'll have to depend on Shoryuken in that case. As of now, I think the current approach will do.

> If we go back with unwrapping the name space it won't be that global anymore.

Good point, I actually never thought of it like that. That said, we still don't use that style and It would add an unwanted constant under `Airbrake::Shoryuken`, too.

phstc replied:
@kyrylo my bad, I forgot to change `def to_s` to `def self.to_s`, now `class.to_s` works fine. 

I guess all feedbacks were addressed then.

kyrylo replied:
Thanks. Don't forget to remove this line :D

kyrylo commented:
I tested this locally and it does work as advertised. Notices seem to be really succinct, though. Would you find any other additional information useful while debugging errors? Maybe queue name? We also have the whole `sqs_msg` object, which carries some info (but I'm not familiar with it). It's ok, if nothing else is useful.

Basically, it needs some minor fixes and then I'll smash that green button :)

phstc commented on spec/unit/shoryuken/error_handler_spec.rb:
```diff
@@ -0,0 +1,64 @@
+RSpec.describe Airbrake::Shoryuken::ErrorHandler do
+  let(:error) { AirbrakeTestError.new('shoryuken error') }
+  let(:body) { { message: 'message' } }
+  let(:notice) { double 'notice' }
```
Great catch! 🍻

phstc commented on spec/unit/shoryuken/error_handler_spec.rb:
```diff
@@ -0,0 +1,64 @@
+        subject.call(worker, nil, nil, body) { raise error }
+      end.to raise_error(error)
+
+      wait_for_a_request_with_body(/"message":"shoryuken\serror"/)
```
> Nitpick: \s isn't needed.

@kyrylo are you suggesting to change `\s` with a whitespace? I copied over the sidekiq example.
 
> This will pass even for the batch test. 

I don't think it would, the line below checks for `body`, the specific test for batch checks for `batch`.

kyrylo replied:
> are you suggesting to change \s with a whitespace? I copied over the sidekiq example.

Yeah, that seems to be unnecessary in this (and Sidekiq's) test.

> I don't think it would, the line below checks for body, the specific test for batch checks for batch.

Good point. Please disregard.

P.S. It's no big deal and it's absolutely not necessary to do this, but ideally we could have 1 spec per 1 `wait_for_a_request_with_body` to avoid duplication of this line.

phstc commented:
> I tested this locally and it does work as advertised. Notices seem to be really succinct, though. Would you find any other additional information useful while debugging errors? Maybe queue name? We also have the whole sqs_msg object, which carries some info (but I'm not familiar with it). It's ok, if nothing else is useful.

@kyrylo I added the queue name, but I don't usually pay attention to it. Usually the worker name is enough. hm... thinking more on it, [I guess requiring Shoryuken will be necessary](https://github.com/airbrake/airbrake/pull/669/files#r101032323), so we can get the proper worker name, WDYT?

> Basically, it needs some minor fixes and then I'll smash that green button :)

Sure, I updated and replied to some.

phstc commented:
> P.S. It's no big deal and it's absolutely not necessary to do this, but
ideally we could have 1 spec per 1 wait_for_a_request_with_body to avoid
duplication of this line.

I know, but then I would have a such hard time to break these regexes
`%r{}` in multiple lines to accommodate the line length limit.

On Tue, Feb 14, 2017 at 8:31 AM, Kyrylo Silin <notifications@github.com>
wrote:

> *@kyrylo* commented on this pull request.
> ------------------------------
>
> In spec/unit/shoryuken/error_handler_spec.rb
> <https://github.com/airbrake/airbrake/pull/669>:
>
> > +
> +  def wait_for_a_request_with_body(body)
> +    wait_for(a_request(:post, endpoint).with(body: body)).to have_been_made.once
> +  end
> +
> +  before do
> +    stub_request(:post, endpoint).to_return(status: 201, body: '{}')
> +  end
> +
> +  context "when there's an error" do
> +    it 'notifies' do
> +      expect do
> +        subject.call(worker, nil, nil, body) { raise error }
> +      end.to raise_error(error)
> +
> +      wait_for_a_request_with_body(/"message":"shoryuken\serror"/)
>
> are you suggesting to change \s with a whitespace? I copied over the
> sidekiq example.
>
> Yeah, that seems to be unnecessary in this (and Sidekiq's) test.
>
> I don't think it would, the line below checks for body, the specific test
> for batch checks for batch.
>
> Good point. Please disregard.
>
> P.S. It's no big deal and it's absolutely not necessary to do this, but
> ideally we could have 1 spec per 1 wait_for_a_request_with_body to avoid
> duplication of this line.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/airbrake/airbrake/pull/669>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGctCoxGVa5tU-adwU2FgPSiTsE0Mulks5rcay9gaJpZM4L9kvj>
> .
>



-- 
Pablo Cantero
pablocantero.com

phstc commented:
Oops! Done :scissors:

On Tue, Feb 14, 2017 at 9:36 AM, Kyrylo Silin <notifications@github.com>
wrote:

> *@kyrylo* commented on this pull request.
> ------------------------------
>
> In spec/unit/shoryuken/error_handler_spec.rb
> <https://github.com/airbrake/airbrake/pull/669>:
>
> > @@ -0,0 +1,64 @@
> +require 'spec_helper'
> +require 'airbrake/shoryuken/error_handler'
> +
> +RSpec.describe Airbrake::Shoryuken::ErrorHandler do
> +  let(:error) { AirbrakeTestError.new('shoryuken error') }
> +  let(:body) { { message: 'message' } }
> +  let(:notice) { double 'notice' }
> +  let(:worker) { FooWorker.new }
> +  let(:endpoint) do
> +    'https://airbrake.io/api/v3/projects/113743/notices?key=fd04e13d806a90f96614ad8e529b2822'
> +  end
> +
> +  class FooWorker; end
>
> Thanks. Don't forget to remove this line :D
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/airbrake/airbrake/pull/669>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGctMHhUIdUtj_42_ldwIRe7TD8xhFNks5rcbvhgaJpZM4L9kvj>
> .
>



-- 
Pablo Cantero
pablocantero.com

kyrylo commented:
>  thinking more on it, I guess requiring Shoryuken will be necessary, so we can get the proper worker name, WDYT?

I'd say this is a job for a separate PR. I'm not sure whether this is mandatory, so let's not bother with this for now :) What we have here is fantastic.

---

Awesome job, this looks ready to me. Merging.
